bentoml/yatai/grpc_stream_utils.py

```diff
@@ -0,0 +1,138 @@
+
+        Args:
+            bento_name:
+            bento_version:
+            directory_path:
+            file_chunk_size:
+            is_request:
+        """
+        self.bento_name = bento_name
+        self.bento_version = bento_version
+        self.directory_path = directory_path
+        self.file_chunk_size = file_chunk_size
+        self.is_request = is_request
+        self.out_stream = FileStream()
+        self.tar = tarfile.TarFile(fileobj=self.out_stream, mode='w')
+
+    @staticmethod
+    def _stream_file_into_tar(tarinfo, tar, file_handler, buf_size):
+        out = tar.fileobj
+
+        # For iter(), the second argument, sentinel, is given, then object must
+        # be a callable object. The iterator created in this case will
+        # call object with no arguments for each call to its __next__()
+        # method; if the value returned is equal to sentinel, StopIteration
+        # will be raised, otherwise the value will be returned.
+        for b in iter(lambda: file_handler.read(buf_size), b''):
+            out.write(b)
+            yield
+        blocks, remainder = divmod(tarinfo.size, tarfile.BLOCKSIZE)
+        if remainder > 0:
+            out.write(tarfile.NUL * (tarfile.BLOCKSIZE - remainder))
+            blocks += 1
+        tar.offset += blocks * tarfile.BLOCKSIZE
+        yield
+
+    def create_request_or_response(self):
+        if self.is_request:
+            return UploadBentoRequest(
+                bento_name=self.bento_name,
+                bento_version=self.bento_version,
+                bento_bundle=self.out_stream.read_value(),
+            )
+        else:
+            return DownloadBentoResponse(bento_bundle=self.out_stream.read_value())
+
+    def close(self):
+        self.tar.close()
+        self.out_stream.close()
+
+    def __iter__(self):
+        if self.is_request:
+            yield UploadBentoRequest(
+                bento_name=self.bento_name, bento_version=self.bento_version
+            )
+        # Include length of separator.
+        prefix_len = len(self.directory_path) + len(os.path.sep)
+
+        # Add the directory path to tar
+        self.tar.add(name=self.directory_path, arcname='', recursive=False)
+
+        # Manually walk the directory and add to tarfile
+        for path, dirs, files in os.walk(self.directory_path):
+            arc_path = path[prefix_len:]
+
+            # Add files to tar. Add the tar info and then stream the file into tar
+            for f in files:
+                file_path = os.path.join(path, f)
+                with open(file_path, 'rb') as file:
+                    # Reading the file info and generate tarinfo from that.
+                    # No file content is added to the tar at this point
+                    tar_info = self.tar.gettarinfo(
+                        name=file_path, arcname=os.path.join(arc_path, f), fileobj=file,
+                    )
+                    self.tar.addfile(tar_info)
```

Comment:
Looking at related docs, `addfile` will 1) add tar info 2) actually write the file to the `tar.fileobj`. whereas `add` will only add the tar info.  In this case we want to stream file content to tar.fileobj in the next step, so I thought it should be `self.tar.add(tar_info)` instead of `self.tar.addfile` here. Could you help me understand this?

Reply:
`tar.add` is typically used for directory, fifo, symbolic link and etc...It generated created a tarinfo and then call `tar.addfile` internally.

We wants to control that process, that's why we are using `tar.addfile`